File: app/src/main/java/com/absinthe/anywhere_/utils/manager/DialogStack.java
package com.absinthe.anywhere_.utils.manager;

import android.app.Dialog;

import androidx.fragment.app.DialogFragment;

import com.absinthe.anywhere_.view.AnywhereBottomSheetDialog;

import java.util.Objects;
import java.util.Stack;

import timber.log.Timber;

/**
 * Created by Absinthe at 2020/1/13
 * <p>
 * Dialog Stack
 * <p>
 * Make it display unique Dialog at the same time
 * to make app delegate.
 */
public class DialogStack {

    public enum Singleton {
        INSTANCE;
        private Stack<Object> instance;

        Singleton() {
            instance = new Stack<>();
        }

        public Stack<Object> getInstance() {
            return instance;
        }
    }

    private static boolean isPrintStack = false;

    public static void push(Object dialog) {
        printStack();
        Timber.i("Push Start");
        if (!(dialog instanceof Dialog) && !(dialog instanceof DialogFragment)) {
            return;
        }

        if (Singleton.INSTANCE.getInstance().empty()) {
            Singleton.INSTANCE.getInstance().push(dialog);
        } else {
            Object peekObject = Singleton.INSTANCE.getInstance().peek();

            if (peekObject instanceof AnywhereBottomSheetDialog) {
                ((AnywhereBottomSheetDialog) peekObject).hide();
            } else if (peekObject instanceof Dialog) {
                ((Dialog) peekObject).hide();
            } else if (peekObject instanceof DialogFragment) {
                Dialog peekDialog = ((DialogFragment) peekObject).getDialog();
                if (peekDialog != null) {
                    peekDialog.hide();
                }
            }
            Singleton.INSTANCE.getInstance().push(dialog);
        }

        if (dialog instanceof AnywhereBottomSheetDialog) {
            ((AnywhereBottomSheetDialog) dialog).isPush = true;
            ((AnywhereBottomSheetDialog) dialog).show();
        } else if (dialog instanceof Dialog) {
            ((Dialog) dialog).show();
        }

        Timber.i("Push End");
        printStack();
    }

    public static void pop() {
        printStack();
        Timber.i("Pop Start");
        if (Singleton.INSTANCE.getInstance().empty()) {
            return;
        }

        Object peekObject = Singleton.INSTANCE.getInstance().peek();

        try {
            if (peekObject instanceof Dialog) {
                ((Dialog) peekObject).dismiss();
            } else if (peekObject instanceof DialogFragment) {
                ((DialogFragment) peekObject).dismiss();
            }

            Singleton.INSTANCE.getInstance().pop();

            if (!Singleton.INSTANCE.getInstance().empty()) {
                peekObject = Singleton.INSTANCE.getInstance().peek();

                if (peekObject == null) {
                    Singleton.INSTANCE.getInstance().pop();
                } else {
                    if (peekObject instanceof Dialog) {
                        ((Dialog) peekObject).show();
                    } else if (peekObject instanceof DialogFragment) {
                        Objects.requireNonNull(
                                ((DialogFragment) peekObject).getDialog()).show();
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        Timber.i("Pop End");
        printStack();
    }

    public static void setPrintStack(boolean flag) {
        isPrintStack = flag;
    }

    private static void printStack() {
        if (isPrintStack) {
            Timber.i("DialogStack:");

            for (Object object : Singleton.INSTANCE.getInstance()) {
                Timber.i(object.getClass().toString());
            }

            Timber.i("--------------------------------------");
        }
    }
}

File: app/src/main/java/com/absinthe/anywhere_/utils/manager/DialogStack.kt
package com.absinthe.anywhere_.utils.manager

import android.app.Dialog
import androidx.fragment.app.DialogFragment
import com.absinthe.anywhere_.view.AnywhereBottomSheetDialog
import timber.log.Timber
import java.util.*

/**
 * Created by Absinthe at 2020/1/13
 *
 *
 * Dialog Stack
 *
 *
 * Make it display unique Dialog at the same time
 * to make app delegate.
 */
object DialogStack {

    private val stack: Stack<Any> = Stack()
    var isPrintStack = false

    fun push(dialog: Any) {
        printStack()
        Timber.i("Push Start")

        if (dialog !is Dialog && dialog !is DialogFragment) {
            return
        }

        if (stack.empty()) {
            stack.push(dialog)
        } else {
            when (val peekObject = stack.peek()) {
                is AnywhereBottomSheetDialog -> {
                    peekObject.hide()
                }
                is Dialog -> {
                    peekObject.hide()
                }
                is DialogFragment -> {
                    peekObject.dialog?.hide()
                }
            }
            stack.push(dialog)
        }

        dialog.apply {
            if (this is AnywhereBottomSheetDialog) {
                isPush = true
                show()
            } else if (this is Dialog) {
                show()
            }
        }

        Timber.i("Push End")
        printStack()
    }

    fun pop() {
        printStack()
        Timber.i("Pop Start")

        if (stack.empty()) {
            return
        }

        stack.peek()?.let {
            try {
                if (it is Dialog) {
                    it.dismiss()
                } else if (it is DialogFragment) {
                    it.dismiss()
                }
                stack.pop()

                if (stack.isNotEmpty()) {
                    val peekObject = stack.peek()

                    if (peekObject == null) {
                        stack.pop()
                    } else {
                        if (peekObject is Dialog) {
                            peekObject.show()
                        } else if (peekObject is DialogFragment) {
                            peekObject.dialog?.show()
                        }
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        Timber.i("Pop End")
        printStack()
    }

    private fun printStack() {
        if (isPrintStack) {
            Timber.i("DialogStack:")

            for (obj in stack) {
                Timber.i(obj.javaClass.toString())
            }
            Timber.i("--------------------------------------")
        }
    }
}